cpp/src/arrow/builder.cc

```diff
@@ -43,6 +44,33 @@ namespace arrow {
 
 using internal::AdaptiveIntBuilderBase;
 
+namespace {
+
+Status TrimBuffer(const int64_t bytes_filled, ResizableBuffer* buffer) {
+  if (buffer) {
+    if (bytes_filled < buffer->size()) {
+      // Trim buffer
+      RETURN_NOT_OK(buffer->Resize(bytes_filled));
+    }
+    // zero the padding
+    buffer->ZeroPadding();
+  } else {
+    DCHECK_EQ(bytes_filled, 0);
```

Comment:
Note null buffers are problematic, see PR #2243.

Reply:
I just left a review + comments there. I think we should try to merge this before merging #2243 so we can decide what to do there